### PR TITLE
ENH: PasteImageFilter input ND and constant value

### PR DESCRIPTION
**Pull request by blowekamp**
Add support to PasteImageFilter to accept a constant value as input,
following the model of the binary function filters.

Add support for the SourceImage input to be a lower dimension than the
destination input. The collapse dimension is specified with the
DestinationSkipAxes parameter.

## PR Checklist
- [ ] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [ ] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)
- [ ] Added test (or behavior not changed)
- [ ] Updated API documentation (or API not changed)
- [ ] Added [license](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Utilities/KWStyle/ITKHeader.h) to new files (if any)
- [ ] Added Python wrapping to new files (if any) as described in [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) Section 9.5
- [ ] Added [ITK examples](https://github.com/InsightSoftwareConsortium/ITKExamples) for all new major features (if any)

Refer to the [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) for
further development details if necessary.

### CODE REVIEW

File: Modules/Filtering/ImageGrid/include/itkPasteImageFilter.h [hunk 1]
```diff
@@ -159,7 +201,10 @@ class ITK_TEMPLATE_EXPORT PasteImageFilter : public InPlaceImageFilter<TInputIma
 
   SourceImageRegionType m_SourceRegion;
 
-  InputImageIndexType m_DestinationIndex;
+  InputImageIndexType    m_DestinationIndex;
+  InputSkipAxesArrayType m_DestinationSkipAxes;
+
+  static_assert(InputImageDimension >= SourceImageDimension, "The source image is greater than the input image.");
```
Comment (dzenanz) [inline Modules/Filtering/ImageGrid/include/itkPasteImageFilter.h:207]: Should `The source image is greater...` possibly be `The source image dimension is greater...`?

File: Modules/Filtering/ImageGrid/include/itkPasteImageFilter.hxx [hunk 2]
```diff
@@ -54,18 +61,57 @@ PasteImageFilter<TInputImage, TSourceImage, TOutputImage>::GenerateInputRequeste
...
+  if (sourceInput == nullptr && constantInput == nullptr)
+  {
+    itkExceptionMacro("The Source or the Constant input are required.")
+  }
+
+
+  auto numberSkippedAxis = std::accumulate(m_DestinationSkipAxes.begin(), m_DestinationSkipAxes.end(), int(0));
+  if (numberSkippedAxis != (InputImageDimension - SourceImageDimension))
+  {
+    itkExceptionMacro("Number of skipped axis " << m_DestinationSkipAxes << " does not match the difference in"
```
Comment (dzenanz) [inline Modules/Filtering/ImageGrid/include/itkPasteImageFilter.hxx:99]: `Number of skipped axis` -> `Number of skipped axes`

File: Modules/Filtering/ImageGrid/include/itkPasteImageFilter.h [hunk 3]
```diff
@@ -20,20 +20,28 @@
...
  *
- * PasteImageFilter allows you to take a section of one image and
- * paste into another image.  The SetDestinationIndex() method
- * prescribes where in the first input to start pasting data from the
- * second input.  The SetSourceRegion method prescribes the section of
- * the second image to paste into the first. If the output requested
- * region does not include the SourceRegion after it has been
- * repositioned to DestinationIndex, then the output will just be
- * a copy of the input.
+ * This filter supports running "InPlace" to efficiently reuses the destination image buffer for the output, removing
```
Comment (zivy) [inline Modules/Filtering/ImageGrid/include/itkPasteImageFilter.h:40]: reuses -> reuse

File: Modules/Filtering/ImageGrid/include/itkPasteImageFilter.h [hunk 4]
```diff
@@ -20,20 +20,28 @@
...
+ * the need to copy the destination pixels to the output.
+ *
+ * When the source image is a lower dimension than the destination image then the DestinationSkipAxes parameter
```
Comment (zivy) [inline Modules/Filtering/ImageGrid/include/itkPasteImageFilter.h:43]: is a lower -> has a lower

File: Modules/Filtering/ImageGrid/include/itkPasteImageFilter.h [hunk 5]
```diff
@@ -101,6 +113,18 @@ class ITK_TEMPLATE_EXPORT PasteImageFilter : public InPlaceImageFilter<TInputIma
   itkSetMacro(DestinationIndex, InputImageIndexType);
   itkGetConstMacro(DestinationIndex, InputImageIndexType);
 
+  /** Set/Get the an array describing which axes in the destination image to skip
```
Comment (zivy) [inline Modules/Filtering/ImageGrid/include/itkPasteImageFilter.h:116]: the an -> the

File: Modules/Filtering/ImageGrid/include/itkPasteImageFilter.hxx [hunk 6]
```diff
@@ -83,59 +129,56 @@ PasteImageFilter<TInputImage, TSourceImage, TOutputImage>::DynamicThreadedGenera
...
-  }
-  else
-  {
-    // Paste region is outside this thread
-    useSource = false;
-  }
+  InputImageRegionType destinationRegion(this->GetDestinationIndex(), destinationSize);
+  // SourceImageRegionType sourceRegionInDestinationImageCropped;
+  // destinationRegion.SetSize(m_SourceRegion.GetSize());
+
```
Comment (zivy) [inline Modules/Filtering/ImageGrid/include/itkPasteImageFilter.hxx:137]: remove the commented out code

File: Modules/Filtering/ImageGrid/include/itkPasteImageFilter.hxx [hunk 7]
```diff
@@ -83,59 +129,56 @@ PasteImageFilter<TInputImage, TSourceImage, TOutputImage>::DynamicThreadedGenera
...
+  // if there is not intersection then the source in not needed
```
Comment (zivy) [inline Modules/Filtering/ImageGrid/include/itkPasteImageFilter.hxx:138]: source in not -> source is not

**Review by dzenanz — APPROVED** [review]
I only glanced over the main logic.

**Review by zivy — COMMENTED** [review]
The pasting goes against the ITK philosophy of "an image is a spatial object", BUT this is where we need to be pragmatists. The functionality is very useful and was missing from the toolkit, so very nice addition to ITK, 👍 .

**Comment by blowekamp** [comment]
Thank you for the review and the recommended improvements.

The reported Mac CI failure is due to a JPEG2000 test failing: https://open.cdash.org/build/6783380 and is not related to this patch.

**Pull request merged**